### PR TITLE
HPCC-29634 Aggregate spill stats to graph/wf scope

### PR DESCRIPTION
**Pull request by shamser**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

**Review by jakesmith — COMMENTED** [review]
I think this looks right.
We discussed this topic in general at our last in-person meeting.

I think this approach is okay for now, but as discussed at the meeting, it could be a little costly if there are many subgraphs, or graphs to aggregate, as it's going to have to go through each blob and fetch, decompress etc.
I suspect it's fine though most of the time, unless there are lots of subgraphs/graphs/wofkflow items at a given level and even then, unless overall they are very quick, the additional aggregation cost I suspect will be negligible compared with the actual runtime.

However, and related to HPCC-29656, I think we agreed the approach should be that the stats should be aggregated all the way up periodically by the engines, so this would change the approach taken here (and HPCC-29656 would no longer be relevant).
NB: This is also related to an older pre-existing JIRA (HPCC-29207).

We also discussed that the in-flight aggregation (for a given level) could be cached (in-memory) whilst in flight, so that it doesn't need to re-fetch each item, if it e.g.. has already aggregated 10 previously complete subgraphs.

For this JIRA/PR I think this approach is sufficient, and we should merge something like.
For HPCC-29656 I suspect it should be closed with a view to implement a scheme like the one discussed.

@ghalliday, can you comment on the PR and the above summary of the approach we discussed at the meeting?

**Comment by shamser** [comment]
> ever, and related to HPCC-29656, I think we agreed the approach should be that the stats should be aggregated all the way up periodically by the engines, so this would change the approach taken here (and HPCC-29656 would no longer be relevant).
> NB: This

I don't think this shouldn't require fetching & decompressing the stats in the stats blob as I've specified "global" as the source.  @ghalliday Please can you confirm if this is the case?

HPCC-29656 may be needed if a work-unit fails and there are aggregates that couldn't be generated.  It should be fast as it checks the aggregates at the global level and if if it's there, it doesn't proceed any further.  Do you think we should have HPCC-29656 in addition to the new  rolling up aggregates mechanism?

**Comment by jakesmith** [comment]
> I don't think this shouldn't require fetching & decompressing the stats in the stats blob as I've specified "global" as the source. @ghalliday Please can you confirm if this is the case?
> 
> HPCC-29656 may be needed if a work-unit fails and there are aggregates that couldn't be generated. It should be fast as it checks the aggregates at the global level and if if it's there, it doesn't proceed any further. Do you think we should have HPCC-29656 in addition to the new rolling up aggregates mechanism?

There will always be a granularity issue where stats. are lost, i.e. thor collects from slaves periodically, so if there's a hard-failure of the master between last collection and hard kill event, then the interim stats. are lost.

If we perform the rollup at these cycles, which is relatively infrequent (between 11 - 19 secs as it stands), then there will nothing else to fill-in/calculate, i.e. all will be rolled up each time stats are gathered from workers.
But we will also need avoid automatically publishing at the end of each sub-graph (HPCC-29207), as that would often be too frequent and costly to Dali/the environment (it already is really, which is why HPCC-29207 was originally opened).
So the new mechanism replaces the need for HPCC-29656

**Review by jakesmith — APPROVED** [review]
> https://github.com/hpcc-systems/HPCC-Platform/pull/17398#pullrequestreview-1498238343

I'd be happy to merge this PR as is, further enhanced + refactoring that was discussed in in-person meetings + related JIRA's, is summarized in the above comment and in sub-task : https://track.hpccsystems.com/browse/HPCC-29657

File: ecl/eclagent/eclagent.cpp [hunk 1]
```diff
@@ -1985,7 +1985,7 @@ void EclAgent::doProcess()
         const cost_type diskAccessCost = aggregateDiskAccessCost(w, nullptr);
         if (diskAccessCost)
             w->setStatistic(queryStatisticsComponentType(), queryStatisticsComponentName(), SSTglobal, "", StCostFileAccess, NULL, diskAccessCost, 1, 0, StatsMergeReplace);
-
+        updateSpillSize(w, nullptr, SSTworkflow);
```
Comment (ghalliday) [inline ecl/eclagent/eclagent.cpp:1988]: Should be SSTglobal.

**Review by ghalliday — COMMENTED** [review]
@shamser looks good execpt for the one correction

**Review by ghalliday — APPROVED** [review]

**Pull request merged**